packages/ckeditor5-markdown-gfm/src/pastefrommarkdownexperimental.ts

```diff
@@ -0,0 +1,195 @@
+	 *
+	 * @private
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _parseMarkdownFromHtml( htmlString: string ): string | null {
+		const withoutOsSpecificTags = this._removeOsSpecificTags( htmlString );
+
+		const withoutWrapperTag = this._isHtmlList( withoutOsSpecificTags ) ?
+			this._removeListWrapperTagsAndBrs( withoutOsSpecificTags ) :
+			this._removeWrapperTag( withoutOsSpecificTags );
+
+		if ( this._containsAnyRemainingHtmlTags( withoutWrapperTag ) ) {
+			return null;
+		}
+
+		return this._replaceHtmlReservedEntitiesWithCharacters( withoutWrapperTag );
+	}
+
+	/**
+	 * Removes OS specific tags.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _removeOsSpecificTags( htmlString: string ): string {
+		// Removing <meta> tag present on Mac.
+		const withoutMetaTag = htmlString.replace( /^<meta\b[^>]*>/, '' ).trim();
+		// Removing <html> tag present on Windows.
+		const withoutHtmlTag = withoutMetaTag.replace( /^<html>/, '' ).replace( /<\/html>$/, '' ).trim();
+		// Removing <body> tag present on Windows.
+		const withoutBodyTag = withoutHtmlTag.replace( /^<body>/, '' ).replace( /<\/body>$/, '' ).trim();
+
+		// Removing <!--StartFragment--> tag present on Windows.
+		return withoutBodyTag.replace( /^<!--StartFragment-->/, '' ).replace( /<!--EndFragment-->$/, '' ).trim();
+	}
+
+	/**
+	 * Removes a single HTML wrapper tag from string.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content without any OS specific tags.
+	 */
+	private _removeWrapperTag( htmlString: string ): string {
+		return htmlString.replace( /^<(\w+)\b[^>]*>\s*([\s\S]*?)\s*<\/\1>/, '$2' ).trim();
+	}
+
+	/**
+	 * Removes multiple HTML wrapper tags from a list of sibling HTML tags.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content without any OS specific tags.
+	 */
+	private _removeListWrapperTagsAndBrs( htmlString: string ): string {
+		const tempDiv = document.createElement( 'div' );
+		tempDiv.innerHTML = htmlString;
+
+		const outerElements = tempDiv.querySelectorAll( ':not(:empty)' );
+		const brElements = tempDiv.querySelectorAll( 'br' );
+
+		for ( const element of outerElements ) {
+			const elementClone = element.cloneNode( true );
+			element.replaceWith( ...elementClone.childNodes );
+		}
+
+		for ( const br of brElements ) {
+			br.replaceWith( '\n' );
+		}
+
+		return tempDiv.innerHTML;
+	}
+
+	/**
+	 * Determines if sting contains sibling HTML tags at root level.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content without any OS specific tags.
+	 */
+	private _isHtmlList( htmlString: string ): boolean {
+		const tempDiv = document.createElement( 'div' );
+		tempDiv.innerHTML = htmlString;
+
+		return tempDiv.children.length > 1;
+	}
+
+	/**
+	 * Determines if string contains any HTML tags.
+	 *
+	 * @private
+	 * @param str
+	 */
+	private _containsAnyRemainingHtmlTags( str: string ): boolean {
+		return /<[^>]+>[\s\S]*<[^>]+>/.test( str );
```

Comment:
Looks like this pattern could be simpler. Wouldn't search for `<` be enough?

Reply:
Good catch.